### PR TITLE
Dev sir 1021 rate limit

### PR DESCRIPTION
**Pull request by OttoVT**
Before merging we need to setup redis in test and change configs

### CODE REVIEW

File: src/Api/RateLimits/RateLimitInterceptor.cs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+                ClientIp = clientIp,
+                Path = httpContext.Method.ToString().ToLowerInvariant(),
+                HttpVerb = "post",
+                ClientId = clientId
+            };
+        }
+
+        protected void LogBlockedRequest(ServerCallContext httpContext, ClientRequestIdentity identity, RateLimitCounter counter, RateLimitRule rule)
+        {
+            _logger.LogInformation($"Request {identity.HttpVerb}:{identity.Path} from ClientId {identity.ClientId} has been blocked, quota {rule.Limit}/{rule.Period} exceeded by {counter.Count}. Blocked by rule {rule.Endpoint}.");
```
Comment (KonstantinRyazantsev) [inline src/Api/RateLimits/RateLimitInterceptor.cs:242]: Use structured logging

File: src/Api/RateLimits/TenantRateLimitMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            IRateLimitConfiguration config,
+            ILogger<TenantRateLimitMiddleware> logger)
+            : base(next, options, new TenantRateLimitProcessor(options, counterStore, policyStore, config), config)
+        {
+            _logger = logger;
+        }
+
+        protected override void LogBlockedRequest(HttpContext httpContext, ClientRequestIdentity identity, RateLimitCounter counter, RateLimitRule rule)
+        {
+            _logger.LogInformation($"Request {identity.HttpVerb}:{identity.Path} from ClientId {identity.ClientId} has been blocked, quota {rule.Limit}/{rule.Period} exceeded by {counter.Count}. Blocked by rule {rule.Endpoint}, TraceIdentifier {httpContext.TraceIdentifier}.");
```
Comment (KonstantinRyazantsev) [inline src/Api/RateLimits/TenantRateLimitMiddleware.cs:25]: use structured logging

File: src/Api/Startup.cs [hunk 3]
```diff
@@ -82,6 +89,47 @@ protected override void ConfigureServicesExt(IServiceCollection services)
...
+        {
+            services.AddSingleton<TenantRateLimitOptions>(Config.TenantRateLimiting);
+            // needed to load configuration from appsettings.json
+            services.AddOptions();
+
+            // needed to store rate limit counters and ip rules
+            //services.AddMemoryCache();
+
+            //load general configuration from appsettings.json
+            //services.Configure<TenantRateLimitOptions>(this.ConfigRoot);
```
Comment (KonstantinRyazantsev) [inline src/Api/Startup.cs:106]: remove it

File: src/Api/Program.cs [hunk 4]
```diff
@@ -36,7 +42,25 @@ public static void Main(string[] args)
...
+                        CancellationToken.None);
+
+                    var get = await cache.GetAsync("test");
+                    //var options = scope.ServiceProvider.GetRequiredService<TenantRateLimitOptions>();
+                    // get the ClientPolicyStore instance
+                    var clientPolicyStore = scope.ServiceProvider.GetRequiredService<IClientPolicyStore>();
+
+                    // seed client data from appsettings
+                    await clientPolicyStore.SeedAsync();
+                }
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Api/Program.cs:61]: oops, that was for test reasons, I will delete it.
await clientPolicyStore.SeedAsync(); - that is needed to populate cache with tenant specific rate limit in the future

**Review by KonstantinRyazantsev — APPROVED** [review]

**Pull request merged**